source/targetfiltersource_test.go

```diff
@@ -0,0 +1,181 @@
+			expected:  []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+		},
+		{
+			title:     "filter matched specific",
+			filters:   endpoint.NewTargetNetFilter([]string{"1.2.3.4"}),
+			endpoints: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+			expected:  []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+		},
+		{
+			title:     "filter matched net",
+			filters:   endpoint.NewTargetNetFilter([]string{"1.2.3.0/24"}),
+			endpoints: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+			expected:  []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+		},
+		{
+			title:     "filter not matched specific",
+			filters:   endpoint.NewTargetNetFilter([]string{"1.2.3.5/32"}),
+			endpoints: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+			expected:  []*endpoint.Endpoint{},
+		},
+		{
+			title:     "filter not matched net",
+			filters:   endpoint.NewTargetNetFilter([]string{"1.2.4.0/24"}),
+			endpoints: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.3.4")},
+			expected:  []*endpoint.Endpoint{},
+		},
+		{
+			title:     "filter not matched CNAME endpoint",
+			filters:   endpoint.NewTargetNetFilter([]string{"1.2.4.0/24"}),
+			endpoints: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "CNAME", "foo.bar")},
+			expected:  []*endpoint.Endpoint{},
+		},
+		{
+			title:   "filter matched one of two",
+			filters: endpoint.NewTargetNetFilter([]string{"1.2.4.0/24"}),
+			endpoints: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("foo", "CNAME", "foo.bar"),
+				endpoint.NewEndpoint("foo", "A", "1.2.4.1")},
+			expected: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "1.2.4.1")},
+		},
+		{
+			title: "filter exclusion all",
+			filters: endpoint.NewTargetNetFilterWithExclusions(
+				[]string{""},
+				[]string{"1.0.0.0/8"}),
+			endpoints: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("foo", "A", "1.2.3.4"),
+				endpoint.NewEndpoint("foo", "A", "1.2.3.5"),
+				endpoint.NewEndpoint("foo", "A", "1.2.3.6"),
+				endpoint.NewEndpoint("foo", "A", "1.3.4.5"),
+				endpoint.NewEndpoint("foo", "A", "1.4.4.5")},
+			expected: []*endpoint.Endpoint{},
+		},
+		{
+			title: "filter exclude internal net",
+			filters: endpoint.NewTargetNetFilterWithExclusions(
+				[]string{""},
+				[]string{"10.0.0.0/8"}),
+			endpoints: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("foo", "A", "10.0.0.1"),
+				endpoint.NewEndpoint("foo", "A", "8.8.8.8")},
+			expected: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "8.8.8.8")},
+		},
+		{
+			title: "filter only internal",
+			filters: endpoint.NewTargetNetFilterWithExclusions(
+				[]string{"10.0.0.0/8"}, []string{}),
+			endpoints: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("foo", "A", "10.0.0.1"),
+				endpoint.NewEndpoint("foo", "A", "8.8.8.8")},
+			expected: []*endpoint.Endpoint{endpoint.NewEndpoint("foo", "A", "10.0.0.1")},
+		},
```

Comment:
This looks like it's testing `endpoint.NewTargetNetFilter()` and `endpoint.NewTargetNetFilterWithExclusions()`. Such tests belong in `endpoint/target_filter_test.go`, not here. This code should limit to the semantics of `source/targetfiltersource.go`.

(If tests are added to `endpoint/target_filter_test.go`, I'd like some IPv6 test cases.)

Reply:
This is testing `targetfiltersource.go`, [see line 174](https://github.com/BadLiveware/external-dns/blob/a55ae004621d7fd7c27ff4985c376b12fe8574ae/source/targetfiltersource_test.go#L174).

The actual bug these test show, and the PR fixes is in `targetfiltersource.go`, [these are the pertinent lines](https://github.com/BadLiveware/external-dns/blob/a55ae004621d7fd7c27ff4985c376b12fe8574ae/source/targetfiltersource.go#L56-L60)

Reply:
There are far more test cases than are needed to test `targetfiltersource.go`. Most of the variants are probing different semantics of `TargetNetFilter`. Those variants don't belong in that test file.

A proper test of `targetfiltersource.go` wouldn't use `TargetNetFilter`; it would use a testing `TargetFilterInterface` which verifies its inputs and returns fixed outputs depending on input and the test case.

Reply:
I see your point and agree, I pushed a fix using a mock implementation of TargetNetFilter